### PR TITLE
[Https] Fix for issue getting private key after having previously exported it

### LINKED ISSUE #21733: [Https] Issue when getting private key after being exported
When running the `dotnet-dev-certs` tool with the `--trust and --export-path --password` options simultaneously, we try to access the private key after the certificate has been exported.

This causes an underlying issue in .NET Core that causes the key to become unusable in future operations, which in turn causes a failure the next time we try to check for a valid certificate so we decide to generate a new one.

Repeating the problem over and over.

The fix will ultimately come from fixing the logic in .NET Core, but in the mean time, we can change our usage in the tool to avoid running into this specific issue.

### PR DESCRIPTION
**Pull request by javiercn**
Fixes https://github.com/dotnet/aspnetcore/issues/21733

The fix here is to cache the results for each certificate we check for exportable keys the first time so as to avoid running into the issue mentioned in .NET Core.

This is a workaround for an underlying runtime issue, which will be fixed in .NET 5.0: https://github.com/dotnet/runtime/issues/36273

### CODE REVIEW

**Comment by Pilchie** [comment]
:eyes:

File: src/Shared/CertificateGeneration/WindowsCertificateManager.cs [hunk 1]
```diff
@@ -10,6 +10,11 @@ internal class WindowsCertificateManager : CertificateManager
     {
         private const int UserCancelledErrorCode = 1223;
 
+        // Cache to store the result of validating whether a key is exportable or not.
+        // This avoid requesting the key multiple times and avoids an issue when we try to access
+        // the key after it has been exported.
+        private Dictionary<string, bool> _areKeysExportable = new Dictionary<string, bool>();
```
Comment (dougbu) [inline src/Shared/CertificateGeneration/WindowsCertificateManager.cs:16]: What prevents this dictionary from unbounded growth?

Comment (pranavkm) [inline src/Shared/CertificateGeneration/WindowsCertificateManager.cs:16]: Does this need to be `OrdinalIgnoreCase` \ `Ordinal`?

File: src/Shared/CertificateGeneration/WindowsCertificateManager.cs [hunk 2]
```diff
@@ -26,10 +31,15 @@ protected override bool IsExportable(X509Certificate2 c)
             // For the first run experience we don't need to know if the certificate can be exported.
             return true;
 #else
-            return (c.GetRSAPrivateKey() is RSACryptoServiceProvider rsaPrivateKey &&
-                    rsaPrivateKey.CspKeyContainerInfo.Exportable) ||
-                (c.GetRSAPrivateKey() is RSACng cngPrivateKey &&
-                    cngPrivateKey.Key.ExportPolicy == CngExportPolicies.AllowExport);
+            if (_areKeysExportable.TryGetValue(c.Thumbprint,out var result))
```
Comment (pranavkm) [inline src/Shared/CertificateGeneration/WindowsCertificateManager.cs:34]:
```suggestion
            if (_areKeysExportable.TryGetValue(c.Thumbprint, out var result))
```

**Review by dougbu — COMMENTED** [review]
If we get dotnet/runtime#36287 into preview5 (not sure why that currently targets 'master'), would this change be needed?

**Review by pranavkm — APPROVED** [review]

File: src/Shared/CertificateGeneration/WindowsCertificateManager.cs (hunk 1 above)
Comment (javiercn) [reply src/Shared/CertificateGeneration/WindowsCertificateManager.cs:16]: > If we get [dotnet/runtime#36287](https://github.com/dotnet/runtime/pull/36287) into preview5 (not sure why that currently targets 'master'), would this change be needed?

From our last tactics discussion dotnet/runtime were planning to do this after preview5. Given that this seems to be happening in preview5, we can pause on this PR.

Comment (mkArtakMSFT) [reply src/Shared/CertificateGeneration/WindowsCertificateManager.cs:16]: I just confirmed with the runtime team - they'll be taking the fix later in 5.0. So this is still needed!

Comment (mkArtakMSFT) [reply src/Shared/CertificateGeneration/WindowsCertificateManager.cs:16]: And it seems I've been overruled by the higher powers and the runtime fix will be brought into 5.0-preview5!

**Pull request closed without merging**